Pebble's not robust to wrong input

If pebble is fed a layer which has a service with no `command` specified, it will panic upon starting that service. The daemon will instead crash immediately upon start if the service's dictionary is empty.

### Steps to reproduce

```shell
git clone https://github.com/canonical/pebble.git
cd pebble
make build
export PEBBLE=`pwd`/pebble
mkdir $PEBBLE/layers
cat >$PEBBLE/layers/001-config.yaml <<EOF
services:
    my-service:
        override: replace
EOF
# NOTE no command on the my-service service
_build/_/bin/pebble run -v --create-dirs
```

In another terminal:

```shell
_build/_/bin/pebble start my-service
```

That command produces the following error in the terminal were the pebble daemon is running:

```
panic: runtime error: slice bounds out of range [1:0]

goroutine 52 [running]:
github.com/canonical/pebble/internal/overlord/servstate.(*ServiceManager).doStart(0xc00014c1e0, 0xc00014b320, 0xc00021c0f0, 0x0, 0x0)
	/home/facu/canonical/sidecar-charm/pebble/internal/overlord/servstate/handlers.go:89 +0x177d
github.com/canonical/pebble/internal/overlord/state.(*TaskRunner).run.func1(0x0, 0x0)
	/home/facu/canonical/sidecar-charm/pebble/internal/overlord/state/taskrunner.go:195 +0xca
gopkg.in/tomb%2ev2.(*Tomb).run(0xc00021c0f0, 0xc000212300)
	/home/facu/go/pkg/mod/gopkg.in/tomb.v2@v2.0.0-20161208151619-d5d1b5820637/tomb.go:163 +0x3c
created by gopkg.in/tomb%2ev2.(*Tomb).Go
	/home/facu/go/pkg/mod/gopkg.in/tomb.v2@v2.0.0-20161208151619-d5d1b5820637/tomb.go:159 +0x112
```

If the service's dictionary is empty:

```
cat >$PEBBLE/layers/001-config.yaml <<EOF
services:
    my-service: ~
EOF
_build/_/bin/pebble run -v --create-dirs
```

That'll produce:

```
2021-06-23T10:21:00.044Z [pebble] Started daemon.
2021/06/23 12:21:00 http: panic serving pid=2083067;uid=1000;socket=/home/facu/canonical/sidecar-charm/pebble/pebble/.pebble.socket;: runtime error: invalid memory address or nil pointer dereference
goroutine 50 [running]:
net/http.(*conn).serve.func1(0xc000208000)
	/usr/lib/go-1.13/src/net/http/server.go:1767 +0x149
panic(0xa032a0, 0xe21550)
	/usr/lib/go-1.13/src/runtime/panic.go:679 +0x1e0
github.com/canonical/pebble/internal/plan.ParseLayer(0x1, 0xc00020208c, 0x6, 0xc000246000, 0x1c, 0x21c, 0x0, 0x0, 0x0)
	/home/facu/canonical/sidecar-charm/pebble/internal/plan/plan.go:279 +0x697
github.com/canonical/pebble/internal/plan.ReadLayersDir(0xc0001fe100, 0x37, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/facu/canonical/sidecar-charm/pebble/internal/plan/plan.go:340 +0xe5d
github.com/canonical/pebble/internal/plan.ReadDir(0xc000028087, 0x30, 0x0, 0x0, 0x0)
	/home/facu/canonical/sidecar-charm/pebble/internal/plan/plan.go:362 +0x27d
github.com/canonical/pebble/internal/overlord/servstate.(*ServiceManager).reloadPlan(0xc000122140, 0x0, 0x0)
	/home/facu/canonical/sidecar-charm/pebble/internal/overlord/servstate/manager.go:60 +0x64
github.com/canonical/pebble/internal/overlord/servstate.(*ServiceManager).acquirePlan(0xc000122140, 0x0, 0x0, 0x0)
	/home/facu/canonical/sidecar-charm/pebble/internal/overlord/servstate/manager.go:174 +0x11e
github.com/canonical/pebble/internal/overlord/servstate.(*ServiceManager).DefaultServiceNames(0xc000122140, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/facu/canonical/sidecar-charm/pebble/internal/overlord/servstate/manager.go:258 +0x7c
github.com/canonical/pebble/internal/daemon.v1PostServices(0xe28520, 0xc00021c200, 0x0, 0x0, 0x0)
	/home/facu/canonical/sidecar-charm/pebble/internal/daemon/api_services.go:73 +0x5f3
github.com/canonical/pebble/internal/daemon.(*Command).ServeHTTP(0xe28520, 0xb2a740, 0xc0002000c0, 0xc00021c200)
	/home/facu/canonical/sidecar-charm/pebble/internal/daemon/daemon.go:248 +0x547
github.com/gorilla/mux.(*Router).ServeHTTP(0xc0001a2000, 0xb2a740, 0xc0002000c0, 0xc00021c200)
	/home/facu/go/pkg/mod/github.com/gorilla/mux@v1.8.0/mux.go:210 +0x1e6
github.com/canonical/pebble/internal/daemon.logit.func1(0xb2ab00, 0xc000222000, 0xc00021c000)
	/home/facu/canonical/sidecar-charm/pebble/internal/daemon/daemon.go:308 +0x143
net/http.HandlerFunc.ServeHTTP(0xc000127200, 0xb2ab00, 0xc000222000, 0xc00021c000)
	/usr/lib/go-1.13/src/net/http/server.go:2007 +0x44
net/http.serverHandler.ServeHTTP(0xc0001b4000, 0xb2ab00, 0xc000222000, 0xc00021c000)
	/usr/lib/go-1.13/src/net/http/server.go:2802 +0x20f
net/http.(*conn).serve(0xc000208000, 0xb2bec0, 0xc0001f4040)
	/usr/lib/go-1.13/src/net/http/server.go:1890 +0x1716
created by net/http.(*Server).Serve
	/usr/lib/go-1.13/src/net/http/server.go:2928 +0x931
```

### Expected behavior

Pebble daemon shouldn't crash, and ideally should return to the client/user some indication that the missing `command` is an issue or that the service's dictionary is invalid.

Also, pebble's README.md lists `command` as an optional field, which apparently it's not.

--- a/README.md
+++ b/README.md
@@ -154,9 +154,9 @@ services:
         # override the existing service spec in the plan with the same name.
         override: merge | replace
 
-        # (Optional) The command to run the service. This is optional in an
-        # individual layer, but required in the combined layer. The command is
-        # executed directly; use "/bin/sh -c '...'" to run via the shell.
+        # (Required in combined layer) The command to run the service. The
+        # command is executed directly; use "/bin/sh -c '...'" to run via the
+        # shell.
         #
         # Example: /usr/bin/somecommand -b -t 30
         command: <commmand>

--- a/README.md
+++ b/README.md
@@ -151,8 +151,15 @@ services:
         #
         # The value 'merge' will ensure that values in this layer specification
         # are merged over existing definitions, whereas 'replace' will entirely
-        # override the existing service spec in the plan with the same name
+        # override the existing service spec in the plan with the same name.
         override: merge | replace
+
+        # (Optional) The command to run the service. This is optional in an
+        # individual layer, but required in the combined layer. The command is
+        # executed directly; use "/bin/sh -c '...'" to run via the shell.
+        #
+        # Example: /usr/bin/somecommand -b -t 30
+        command: <commmand>
 
         # (Optional) A short summary of the service
         summary: <summary>
@@ -160,10 +167,6 @@ services:
         # (Optional) A detailed description of the service
         description: |
             <description>
-
-        # (Optional) The command to run the service.
-        # Example: /usr/bin/somecommand -b -t 30
-        command: <commmand>
 
         # (Optional) Control whether the service is started automatically when
         # Pebble starts. 

--- a/internal/plan/plan.go
+++ b/internal/plan/plan.go
@@ -83,7 +83,7 @@ func (e *FormatError) Error() string {
 	return e.Message
 }
 
-// CombineLayers combines the given layers into a plan, with the later layers
+// CombineLayers combines the given layers into a single layer, with the later
 // layers overriding earlier ones.
 func CombineLayers(layers ...*Layer) (*Layer, error) {
 	combined := &Layer{

--- a/internal/plan/plan.go
+++ b/internal/plan/plan.go
@@ -150,13 +150,16 @@ func CombineLayers(layers ...*Layer) (*Layer, error) {
 	}
 
 	// Ensure fields in combined layers validate correctly.
-	err := combined.validateFields()
-	if err != nil {
-		return nil, err
+	for name, service := range combined.Services {
+		if service.Command == "" {
+			return nil, &FormatError{
+				Message: fmt.Sprintf(`plan must define "command" for service %q`, name),
+			}
+		}
 	}
 
 	// Ensure combined layers don't have cycles.
-	err = combined.checkCycles()
+	err := combined.checkCycles()
 	if err != nil {
 		return nil, err
 	}
@@ -254,17 +257,6 @@ func order(services map[string]*Service, names []string, stop bool) ([]string, e
 	return order, nil
 }
 
-func (l *Layer) validateFields() error {
-	for name, service := range l.Services {
-		if service.Command == "" {
-			return &FormatError{
-				Message: fmt.Sprintf(`plan must define "command" for service %q`, name),
-			}
-		}
-	}
-	return nil
-}
-
 func (l *Layer) checkCycles() error {
 	var names []string
 	for name := range l.Services {
@@ -287,6 +279,11 @@ func ParseLayer(order int, label string, data []byte) (*Layer, error) {
 	layer.Order = order
 	layer.Label = label
 	for name, service := range layer.Services {
+		if name == "" {
+			return nil, &FormatError{
+				Message: fmt.Sprintf("cannot use empty string as service name"),
+			}
+		}
 		if name == "pebble" {
 			// Disallow service name "pebble" to avoid ambiguity (for example,
 			// in log output).

--- a/internal/plan/plan.go
+++ b/internal/plan/plan.go
@@ -83,7 +83,7 @@ func (e *FormatError) Error() string {
 	return e.Message
 }
 
-// CombineLayers combines the given layers into a Plan, with the later layers
+// CombineLayers combines the given layers into a plan, with the later layers
 // layers overriding earlier ones.
 func CombineLayers(layers ...*Layer) (*Layer, error) {
 	combined := &Layer{
@@ -149,11 +149,18 @@ func CombineLayers(layers ...*Layer) (*Layer, error) {
 		}
 	}
 
-	// Ensure combined layers don't have cycles.
-	err := combined.checkCycles()
+	// Ensure fields in combined layers validate correctly.
+	err := combined.validateFields()
 	if err != nil {
 		return nil, err
 	}
+
+	// Ensure combined layers don't have cycles.
+	err = combined.checkCycles()
+	if err != nil {
+		return nil, err
+	}
+
 	return combined, nil
 }
 
@@ -247,6 +254,17 @@ func order(services map[string]*Service, names []string, stop bool) ([]string, e
 	return order, nil
 }
 
+func (l *Layer) validateFields() error {
+	for name, service := range l.Services {
+		if service.Command == "" {
+			return &FormatError{
+				Message: fmt.Sprintf(`plan must define "command" for service %q`, name),
+			}
+		}
+	}
+	return nil
+}
+
 func (l *Layer) checkCycles() error {
 	var names []string
 	for name := range l.Services {
@@ -274,6 +292,11 @@ func ParseLayer(order int, label string, data []byte) (*Layer, error) {
 			// in log output).
 			return nil, &FormatError{
 				Message: fmt.Sprintf("cannot use reserved service name %q", name),
+			}
+		}
+		if service == nil {
+			return nil, &FormatError{
+				Message: fmt.Sprintf("service object cannot be null for service %q", name),
 			}
 		}
 		service.Name = name

--- a/internal/plan/plan_test.go
+++ b/internal/plan/plan_test.go
@@ -327,6 +327,15 @@ var planTests = []planTest{{
 		services:
 			svc1: ~
 	`},
+}, {
+	summary: `Cannot use empty string as service name`,
+	error:   "cannot use empty string as service name",
+	input: []string{`
+		services:
+			"":
+				override: replace
+				command: cmd
+	`},
 }}
 
 func (s *S) TestParseLayer(c *C) {

--- a/internal/plan/plan_test.go
+++ b/internal/plan/plan_test.go
@@ -320,6 +320,13 @@ var planTests = []planTest{{
 			pebble:
 				command: cmd
 	`},
+}, {
+	summary: `Cannot have null service definition`,
+	error:   `service object cannot be null for service "svc1"`,
+	input: []string{`
+		services:
+			svc1: ~
+	`},
 }}
 
 func (s *S) TestParseLayer(c *C) {
@@ -401,10 +408,45 @@ services:
     srv1:
         command: cmd
 `))
+	c.Assert(err, IsNil)
 	_, err = plan.CombineLayers(layer1, layer2)
 	c.Check(err, ErrorMatches, `layer "label2" must define \"override\" for service "srv1"`)
 	_, ok := err.(*plan.FormatError)
 	c.Check(ok, Equals, true, Commentf("error must be *plan.FormatError, not %T", err))
+}
+
+func (s *S) TestMissingCommand(c *C) {
+	// Combine fails if no command in combined plan
+	layer1, err := plan.ParseLayer(1, "label1", []byte("{}"))
+	c.Assert(err, IsNil)
+	layer2, err := plan.ParseLayer(2, "label2", []byte(`
+services:
+    srv1:
+        override: merge
+`))
+	c.Assert(err, IsNil)
+	_, err = plan.CombineLayers(layer1, layer2)
+	c.Check(err, ErrorMatches, `plan must define "command" for service "srv1"`)
+	_, ok := err.(*plan.FormatError)
+	c.Check(ok, Equals, true, Commentf("error must be *plan.FormatError, not %T", err))
+
+	// Combine succeeds if there is a command in combined plan
+	layer1, err = plan.ParseLayer(1, "label1", []byte(`
+services:
+    srv1:
+        override: merge
+        command: foo --bar
+`))
+	c.Assert(err, IsNil)
+	layer2, err = plan.ParseLayer(2, "label2", []byte(`
+services:
+    srv1:
+        override: merge
+`))
+	c.Assert(err, IsNil)
+	combined, err := plan.CombineLayers(layer1, layer2)
+	c.Assert(err, IsNil)
+	c.Assert(combined.Services["srv1"].Command, Equals, "foo --bar")
 }
 
 func (s *S) TestReadDir(c *C) {

